File: 백준/Gold/16234. 인구 이동/인구 이동.py
from  collections import deque
import copy,sys
input = sys.stdin.readline

def bfs(ny,nx):
    q = deque()
    q.append((ny,nx))
    people = set()
    people.add((ny,nx))
    while q:
        y,x = q.popleft()

        for k in range(4):
            dy = diy[k] + y
            dx = dix[k] + x
            if dy < 0 or dy >= N or dx < 0 or dx >= N:
                continue
            if border[dy][dx] == 1:
                continue
            if (dy,dx) in people:
                continue
            if L <= abs(arr[y][x] - arr[dy][dx]) <= R:
                border[dy][dx] = 1
                q.append((dy,dx))
                people.add((dy,dx))

    #인구이동
    sum_people = 0
    for y,x in people:
        sum_people += arr[y][x]
    avg_people = int(sum_people / len(people))
    if avg_people == sum_people:
        return
    for y,x in people:
        backup_arr[y][x] = avg_people


N, L, R = map(int,input().split())
arr = []
diy = [0,0,1,-1]
dix = [1,-1,0,0]
ans = 0

for _ in range(N):
    arr.append(list(map(int, input().split())))

while 1:
    border = [[0] * N for _ in range(N)]
    backup_arr = copy.deepcopy(arr)
    # 국경체크
    for i in range(N):
        for j in range(N):
            if border[i][j] ==0:
                border[i][j] = 1
                bfs(i,j)

    if arr == backup_arr:
        break
    arr = copy.deepcopy(backup_arr)
    ans += 1

print(ans)

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, L, R;
    static int[][] arr;
    static int[] dy = {0, 0, 1, -1};
    static int[] dx = {1, -1, 0, 0};

    static int[][] border;
    static int[][] backup;
    static boolean moved;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] nlr = br.readLine().split(" ");
        N = Integer.parseInt(nlr[0]);
        L = Integer.parseInt(nlr[1]);
        R = Integer.parseInt(nlr[2]);

        arr = new int[N][N];
        for (int i = 0; i < N; i++) {
            String[] row = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(row[j]);
            }
        }

        int ans = 0;

        while (true) {
            border = new int[N][N];
            backup = new int[N][N];
            for (int i = 0; i < N; i++) {
                backup[i] = Arrays.copyOf(arr[i], N);
            }

            moved = false;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (border[i][j] == 0) {
                        bfs(i, j);
                    }
                }
            }

            if (!moved) break;

            for (int i = 0; i < N; i++) {
                arr[i] = Arrays.copyOf(backup[i], N);
            }

            ans++;
        }

        System.out.println(ans);
    }

    static void bfs(int sy, int sx) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{sy, sx});

        Set<String> people = new HashSet<>();
        people.add(sy + "," + sx);

        border[sy][sx] = 1;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int y = cur[0], x = cur[1];

            for (int d = 0; d < 4; d++) {
                int ny = y + dy[d];
                int nx = x + dx[d];

                if (ny < 0 || nx < 0 || ny >= N || nx >= N) continue;
                if (border[ny][nx] == 1) continue;

                int diff = Math.abs(arr[y][x] - arr[ny][nx]);
                if (L <= diff && diff <= R) {
                    border[ny][nx] = 1;
                    q.offer(new int[]{ny, nx});
                    people.add(ny + "," + nx);
                }
            }
        }

        if (people.size() == 1) return;

        moved = true;
        int total = 0;
        for (String pos : people) {
            String[] parts = pos.split(",");
            int y = Integer.parseInt(parts[0]);
            int x = Integer.parseInt(parts[1]);
            total += arr[y][x];
        }

        int avg = total / people.size();

        for (String pos : people) {
            String[] parts = pos.split(",");
            int y = Integer.parseInt(parts[0]);
            int x = Integer.parseInt(parts[1]);
            backup[y][x] = avg;
        }
    }
}
